java/src/main/native/ai_onnxruntime_OrtTrainingSession.c

```diff
@@ -0,0 +1,745 @@
+  (void) clazz; // Required JNI parameters not needed by functions which don't need to access their host class.
+  const OrtApi* api = (const OrtApi*) apiHandle;
+  const OrtTrainingApi* trainApi = (const OrtTrainingApi*) trainApiHandle;
+  const OrtEnv* env = (const OrtEnv*) envHandle;
+  const OrtSessionOptions* options = (const OrtSessionOptions*) optionsHandle;
+  OrtCheckpointState* checkpoint = (OrtCheckpointState*) checkpointHandle;
+
+  jsize evalLength = (*jniEnv)->GetStringLength(jniEnv, evalPath);
+  jsize optimizerLength = (*jniEnv)->GetStringLength(jniEnv, optimizerPath);
+  OrtTrainingSession* session = NULL;
+
+#ifdef _WIN32
+  // The output of GetStringChars is not null-terminated, so we copy it and add a terminator
+  const jchar* trainJavaStr = (*jniEnv)->GetStringChars(jniEnv, trainPath, NULL);
+  size_t trainStrLength = (*jniEnv)->GetStringLength(jniEnv, trainPath);
+  wchar_t* trainStr = (wchar_t*)calloc(trainStrLength + 1, sizeof(wchar_t));
+  if (trainStr == NULL) {
+    (*jniEnv)->ReleaseStringChars(jniEnv, trainPath, trainJavaStr);
+    throwOrtException(jniEnv, 1, "Not enough memory");
+    return (jlong) session;
+  }
+  const jchar* evalJavaStr = (*jniEnv)->GetStringChars(jniEnv, evalPath, NULL);
+  size_t evalStrLength = (*jniEnv)->GetStringLength(jniEnv, evalPath);
+  wchar_t* evalStr = (wchar_t*)calloc(evalStrLength + 1, sizeof(wchar_t));
+  if (evalStr == NULL) {
+    (*jniEnv)->ReleaseStringChars(jniEnv, trainPath, trainJavaStr);
+    (*jniEnv)->ReleaseStringChars(jniEnv, evalPath, evalJavaStr);
+    free(trainStr);
+    throwOrtException(jniEnv, 1, "Not enough memory");
+    return (jlong) session;
+  }
+  const jchar* optimizerJavaStr = (*jniEnv)->GetStringChars(jniEnv, optimizerPath, NULL);
+  size_t optimizerStrLength = (*jniEnv)->GetStringLength(jniEnv, optimizerPath);
+  wchar_t* optimizerStr = (wchar_t*)calloc(optimizerStrLength + 1, sizeof(wchar_t));
+  if (optimizerStr == NULL) {
+    (*jniEnv)->ReleaseStringChars(jniEnv, trainPath, trainJavaStr);
+    (*jniEnv)->ReleaseStringChars(jniEnv, evalPath, evalJavaStr);
+    (*jniEnv)->ReleaseStringChars(jniEnv, optimizerPath, optimizerJavaStr);
+    free(trainStr);
+    free(evalStr);
+    throwOrtException(jniEnv, 1, "Not enough memory");
+    return (jlong) session;
+  }
+  wcsncpy_s(trainStr, trainStrLength + 1, (const wchar_t*)trainJavaStr, trainStrLength);
+  wcsncpy_s(evalStr, evalStrLength + 1, (const wchar_t*)evalJavaStr, evalStrLength);
+  wcsncpy_s(optimizerStr, optimizerStrLength + 1, (const wchar_t*)optimizerJavaStr, optimizerStrLength);
+  (*jniEnv)->ReleaseStringChars(jniEnv, trainPath, trainJavaStr);
+  (*jniEnv)->ReleaseStringChars(jniEnv, evalPath, evalJavaStr);
+  (*jniEnv)->ReleaseStringChars(jniEnv, optimizerPath, optimizerJavaStr);
+  checkOrtStatus(jniEnv, api, trainApi->CreateTrainingSession(env, options, checkpoint, trainStr, evalStr, optimizerStr, &session));
+  free(trainStr);
+  free(evalStr);
+  free(optimizerStr);
```

Comment:
This release section is growing and duplicaterd. Would you not rather use goto: to the specific cleanup at the end?

Reply:
I think I'll split it into two methods and use the ifdef to dispatch to them, then use gotos in the windows case. I didn't like the combination of goto and ifdef I had when I was developing it as it was a bit unpleasant to reason over.

Reply:
I refactored it further as NULL is used as a signalling mechanism for the underlying CreateTrainingSession call, and the Windows logic was broken in that case.